### ISSUE TITLE
[java] OBL_UNSATISFIED_OBLIGATION / ODR_OPEN_DATABASE_RESOURCE missed when variable is declared as supertype Statement

### ISSUE THREAD

**Issue by Emilyaxe** [spotbugs/spotbugs#4019]
**Describe the bug**
SpotBugs reports a resource leak when a `PreparedStatement` variable is not closed, but misses the same leak when the variable is declared as its supertype `Statement` and assigned from `conn.prepareStatement(...)`. 

Both cases create the same `PreparedStatement` resource and neither closes it, so the second case should also trigger `OBL_UNSATISFIED_OBLIGATION` / `ODR_OPEN_DATABASE_RESOURCE`. This results in a false negative.

**To Reproduce**

```java
import java.sql.*;

public class ResourceLeakDemo {
    static final String DB_URL = "jdbc:h2:mem:testdb";

    // BEFORE: flagged (OBL_UNSATISFIED_OBLIGATION, ODR_OPEN_DATABASE_RESOURCE)
    public void fetchUserBefore(int id) throws SQLException {
        Connection conn = null;
        PreparedStatement ps = null;   // declared as PreparedStatement
        try {
            conn = DriverManager.getConnection(DB_URL);
            ps = conn.prepareStatement("SELECT * FROM users WHERE id = ?");
            ps.setInt(1, id);
            ps.executeQuery();
        } finally {
            // ps intentionally NOT closed
            if (conn != null) conn.close();  
        }
    }

    // AFTER: no warning, but leak is identical
    public void fetchUserAfter(int id) throws SQLException {
        Connection conn = null;
        Statement ps = null;           // declared as supertype Statement
        try {
            conn = DriverManager.getConnection(DB_URL);
            ps = conn.prepareStatement("SELECT * FROM users WHERE id = ?");
            ((PreparedStatement) ps).setInt(1, id);
            ps.executeQuery();
        } finally {
            // ps intentionally NOT closed   // false negative
            if (conn != null) conn.close();  
        }
    }
}
```

Run with:

```
spotbugs -textui ResourceLeakDemo.class
```

**Expected behavior**
Both methods should be reported — the unclosed JDBC statement exists in both, regardless of whether the local variable's declared type is `PreparedStatement` or `Statement`.

**Environment**

- Java version: JDK 21 (Eclipse Temurin)
- SpotBugs Version: 4.9.8
- How do you use SpotBugs: CLI
